File: src/domain/models/detection_result.py
"""DetectionResult value object and ConflictSeverity enum.

Captures the outcome of a tool detection scan: which tools were found,
any configuration conflicts, and their severity classification.
"""

from __future__ import annotations

import enum
from dataclasses import dataclass, field
from types import MappingProxyType
from typing import TYPE_CHECKING

if TYPE_CHECKING:
    from collections.abc import Mapping

    from src.domain.models.detected_tool import DetectedTool


class ConflictSeverity(enum.Enum):
    """Severity of a configuration conflict between global and local settings.

    COMPATIBLE: Global and local settings agree; no action needed.
    WARNING:    A potential issue that the user should be aware of.
    CONFLICT:   A direct contradiction that may cause problems.
    """

    COMPATIBLE = "compatible"
    WARNING = "warning"
    CONFLICT = "conflict"


@dataclass(frozen=True)
class DetectionResult:
    """Immutable result of scanning for AI coding tools and conflicts.

    Attributes:
        detected_tools: Tuple of detected tool value objects.
        conflicts: Tuple of human-readable conflict descriptions.
        severity_map: Read-only mapping from conflict description to severity.
    """

    detected_tools: tuple[DetectedTool, ...]
    conflicts: tuple[str, ...]
    severity_map: Mapping[str, ConflictSeverity] = field(default_factory=dict)

    def __post_init__(self) -> None:
        """Wrap severity_map in MappingProxyType to enforce immutability."""
        object.__setattr__(self, "severity_map", MappingProxyType(dict(self.severity_map)))

File: internal/discovery/domain/detection_result.go
package domain

// ConflictSeverity classifies the severity of a configuration conflict.
type ConflictSeverity string

// ConflictSeverity constants.
const (
	SeverityCompatible ConflictSeverity = "compatible"
	SeverityWarning    ConflictSeverity = "warning"
	SeverityConflict   ConflictSeverity = "conflict"
)

// AllConflictSeverities returns all valid severity values.
func AllConflictSeverities() []ConflictSeverity {
	return []ConflictSeverity{SeverityCompatible, SeverityWarning, SeverityConflict}
}

// DetectionResult is an immutable result of scanning for AI coding tools and conflicts.
type DetectionResult struct {
	severityMap   map[string]ConflictSeverity
	detectedTools []DetectedTool
	conflicts     []string
}

// NewDetectionResult creates a DetectionResult with defensive copies.
func NewDetectionResult(
	detectedTools []DetectedTool,
	conflicts []string,
	severityMap map[string]ConflictSeverity,
) DetectionResult {
	dt := make([]DetectedTool, len(detectedTools))
	copy(dt, detectedTools)
	c := make([]string, len(conflicts))
	copy(c, conflicts)
	sm := make(map[string]ConflictSeverity, len(severityMap))
	for k, v := range severityMap {
		sm[k] = v
	}
	return DetectionResult{detectedTools: dt, conflicts: c, severityMap: sm}
}

// DetectedTools returns a defensive copy of detected tools.
func (dr DetectionResult) DetectedTools() []DetectedTool {
	out := make([]DetectedTool, len(dr.detectedTools))
	copy(out, dr.detectedTools)
	return out
}

// Conflicts returns a defensive copy of conflict descriptions.
func (dr DetectionResult) Conflicts() []string {
	out := make([]string, len(dr.conflicts))
	copy(out, dr.conflicts)
	return out
}

// SeverityMap returns a defensive copy of the severity mapping.
func (dr DetectionResult) SeverityMap() map[string]ConflictSeverity {
	out := make(map[string]ConflictSeverity, len(dr.severityMap))
	for k, v := range dr.severityMap {
		out[k] = v
	}
	return out
}
